pkg/image/admission/imagepolicy/rules/accept.go

```diff
@@ -0,0 +1,133 @@
+package rules
+
+import (
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+
+	"github.com/openshift/origin/pkg/image/admission/imagepolicy/api"
+)
+
+type Accepter interface {
+	Covers(unversioned.GroupResource) bool
+	RequiresImage(unversioned.GroupResource) bool
+	ResolvesImage(unversioned.GroupResource) bool
+
+	Accepts(*ImagePolicyAttributes) bool
+}
+
+// mappedAccepter implements the Accepter interface for a map of group resources and accepters
+type mappedAccepter map[unversioned.GroupResource]Accepter
+
+func (a mappedAccepter) Covers(gr unversioned.GroupResource) bool {
+	_, ok := a[gr]
+	return ok
+}
+
+func (a mappedAccepter) RequiresImage(gr unversioned.GroupResource) bool {
+	accepter, ok := a[gr]
+	return ok && accepter.RequiresImage(gr)
+}
+
+func (a mappedAccepter) ResolvesImage(gr unversioned.GroupResource) bool {
+	accepter, ok := a[gr]
+	return ok && accepter.ResolvesImage(gr)
+}
+
+// Accepts returns true if no Accepter is registered for the group resource in attributes,
+// or if the registered Accepter also returns true.
+func (a mappedAccepter) Accepts(attr *ImagePolicyAttributes) bool {
+	accepter, ok := a[attr.Resource]
+	if !ok {
+		return true
+	}
+	return accepter.Accepts(attr)
+}
+
+type executionAccepter struct {
+	rules         []api.ImageExecutionPolicyRule
+	covers        unversioned.GroupResource
+	defaultReject bool
+	requiresImage bool
+	resolvesImage bool
+
+	integratedRegistryMatcher RegistryMatcher
+}
+
+// NewExecutionRuleseAccepter creates an Accepter from the provided rules.
+func NewExecutionRulesAccepter(rules []api.ImageExecutionPolicyRule, integratedRegistryMatcher RegistryMatcher) Accepter {
+	mapped := make(mappedAccepter)
+
+	for _, rule := range rules {
+		requiresImage, over := prepareSourceRule(&rule.ImageCondition)
+		for gr := range over {
+			a, ok := mapped[gr]
+			if !ok {
+				a = &executionAccepter{
+					covers: gr,
+					integratedRegistryMatcher: integratedRegistryMatcher,
+				}
+				mapped[gr] = a
+			}
+			byResource := a.(*executionAccepter)
+			byResource.rules = append(byResource.rules, rule)
+			if rule.Resolve {
+				byResource.resolvesImage = true
+			}
+			if requiresImage || rule.Resolve {
+				byResource.requiresImage = true
+			}
+		}
+	}
+
+	for _, a := range mapped {
+		byResource := a.(*executionAccepter)
+		if len(byResource.rules) > 0 {
+			// if all rules are reject, the default behavior is allow
+			allReject := true
+			for _, rule := range byResource.rules {
+				if !rule.Reject {
+					allReject = false
+					break
+				}
+			}
+			byResource.defaultReject = !allReject
+		}
+	}
+
+	return mapped
+}
+
+func (r *executionAccepter) RequiresImage(gr unversioned.GroupResource) bool {
+	return r.requiresImage && r.Covers(gr)
```

Comment:
RequiresImage cannot be true if Covers is false.  The API allows the rule
acceptor to lazily load the necessary data - and this should be a zero
allocation pass.

On Fri, Aug 5, 2016 at 9:24 AM, David Eads <notifications@github.com> wrote:

> In pkg/image/admission/imagepolicy/rules/accept.go
> <https://github.com/openshift/origin/pull/8995#discussion_r73691006>:
>
> > +			allReject := true
> > +			for _, rule := range byResource.rules {
> > +				if !rule.Reject {
> > +					allReject = false
> > +					break
> > +				}
> > +			}
> > +			byResource.defaultReject = !allReject
> > +		}
> > +	}
> > +
> > +	return mapped
> > +}
> > +
> > +func (r *executionAccepter) RequiresImage(gr unversioned.GroupResource) bool {
> > +	return r.requiresImage && r.Covers(gr)
>
> I'd rather test .Covers separately.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/8995/files/f96f5e56ba70c7c0f24f875621efdf1ef45ce3c0#r73691006>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p75AoeCt2S4gYyKp4n-M1ZDKjn1iks5qczmDgaJpZM4IlDlb>
> .
>